### PR TITLE
Create the asset list view

### PR DESCRIPTION
**Pull request by udayasri**
Add TableViewDataSource class to make the tableview data source & model agnostic
Fetch Assets from the Exposure api
Add UITableView+EmptyMessage to show empty message when tableview has no data
Handle logout the user from the application

### CODE REVIEW

File: RefApp/ViewControllers/AssetListTableViewController.swift [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+//
+//  Created by Udaya Sri Senarathne on 2018-11-21.
+//  Copyright © 2018 emp. All rights reserved.
+//
+
+import UIKit
+import Exposure
+import ExposurePlayback
+
+class AssetListTableViewController: UITableViewController {
```
Comment (FredrikSjoberg) [inline RefApp/ViewControllers/AssetListTableViewController.swift]: The architecture implies a new viewController for other "types" of lists.
* epg
* asset types
* etc..

File: RefApp/ViewControllers/AssetListTableViewController.swift [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+        self.popupAlert(title: title, message: message, actions: [logOutAction, cancelAction])
+    }
+    
+    /// Log out the user from the application
+    fileprivate func logoutUser() {
+        
+        let navigationController = MainNavigationController()
+        
+        guard let environment = StorageProvider.storedEnvironment, let sessionToken = StorageProvider.storedSessionToken else {
+
```
Comment (FredrikSjoberg) [inline RefApp/ViewControllers/AssetListTableViewController.swift]: Logging out will present a new `MainNavigationController` modally which in turn will push an `EnvironmentViewController` to the stack. If the user proceeds to select env and then log in, another `MainNavigationController` will be presented modally by the parent view controller which in this case is the original `MainNavigationController`...


Repeating this procedure will allow a user to create an ever increasing stack of modally presented `MainNavigationController`s.

```
MainNavigationController -> MainNavigationController -> MainNavigationController -> MainNavigationController
```

I would suggest using push/pop for the relevant transitions, or if modal presentation is desired, pair each `present(...)` with a `dismiss(...)`

File: RefApp/ViewControllers/LoginViewController.swift [hunk 3]
```diff
@@ -170,6 +170,9 @@ extension LoginViewController {
                 if let credentials = $0.value {
                     StorageProvider.store(environment: self?.environment)
                     StorageProvider.store(sessionToken: credentials.sessionToken)
+                    
+                    let navigationController = MainNavigationController()
```
Comment (FredrikSjoberg) [inline RefApp/ViewControllers/LoginViewController.swift]: See comment above about `MainNavigationController` stacks

File: RefApp/ViewControllers/AssetListTableViewController.swift [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+        self.navigationItem.leftBarButtonItem = barButton
+    }
+}
+
+// MARK: - DataSource
+extension AssetListTableViewController {
+    
+    /// Generate tableview content by loading assets from API
+    fileprivate func generateTableViewContent() {
+        guard let environment = StorageProvider.storedEnvironment, let _ = StorageProvider.storedSessionToken else {
```
Comment (FredrikSjoberg) [inline RefApp/ViewControllers/AssetListTableViewController.swift]: A session token is not required to fetch assets but the guard statement indicates this is the case.

I would prefer checking the sessionToken once it is required, ie when making a play call. If no token is available then, prompt the user to login before making the play request

File: RefApp/Utilities/TableViewDataSource.swift [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+        } else {
+            tableView.hideEmptyMessage()
+        }
+        return models.count
+    }
+    
+    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        let model = models[indexPath.row]
+        let cell = tableView.dequeueReusableCell(withIdentifier: reuseIdentifier, for: indexPath)
+        cellConfigurator(model, cell)
```
Comment (FredrikSjoberg) [inline RefApp/Utilities/TableViewDataSource.swift]: Nothing major in this context but what is the entire purpose of `cellConfugurator(...)`?

Just setting cell appearance from already known data?

If so, I suggest using 
`func tableView(_ tableView: UITableView, willDisplay cell: UITableViewCell, forRowAt indexPath: IndexPath)` to customise the appearance of the cell as it arrives.

That would of cause complicate things by making `TableViewDataSource` both the tableview's datasource and delegate

**Comment by FredrikSjoberg** [comment]
Also, we need a deployment target of 9.0

This will conflict with certain properties and functions such as
`navigationBar.prefersLargeTitles = true`

**Comment by FredrikSjoberg** [comment]
One final thing I forgot, it would look good if we add a matching dark background color to the ListViewController and its cells and switch cell titles to white

File: RefApp/ViewControllers/AssetListTableViewController.swift [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+            logoutUser()
+            return
+        }
+        
+        let query = "assetType=MOVIE"
+        DispatchQueue.global(qos: .userInitiated).async { [weak self] in
```
[the remark this answers is not in the document]
Comment (udayasri) [reply RefApp/ViewControllers/AssetListTableViewController.swift]: Ok nice

File: RefApp/ViewControllers/AssetListTableViewController.swift [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+        }
+    }
+    
+    
+    /// Create the datasource for the tableview
+    ///
+    /// - Parameter assets: loaded assets from Exposure API
+    func assetsDidLoad(_ assets: [Asset]) {
+        datasource = .make(for: assets)
+        DispatchQueue.main.async {
```
[the remark this answers is not in the document]
Comment (udayasri) [reply RefApp/ViewControllers/AssetListTableViewController.swift]: 👍

File: RefApp/ViewControllers/AssetListTableViewController.swift (hunk 2 above)
Comment (udayasri) [reply RefApp/ViewControllers/AssetListTableViewController.swift]: yes, totally agree. stupid mistake. will use push & pop

File: RefApp/ViewControllers/LoginViewController.swift (hunk 3 above)
Comment (udayasri) [reply RefApp/ViewControllers/LoginViewController.swift]: 👍

File: RefApp/ViewControllers/AssetListTableViewController.swift (hunk 4 above)
Comment (udayasri) [reply RefApp/ViewControllers/AssetListTableViewController.swift]: I thought it is good to check the session token before fetching assets. But may be as it is not needed to fetch the assets, there is no point of checking for the token. Will remove it from the guard statement.

File: RefApp/ViewControllers/AssetListTableViewController.swift (hunk 1 above)
Comment (FredrikSjoberg) [reply RefApp/ViewControllers/AssetListTableViewController.swift]: I don't think its necessary to make any changes for this PR, but could be something to consider for the future

Comment (udayasri) [reply RefApp/ViewControllers/AssetListTableViewController.swift]: ok

File: RefApp/ViewControllers/AssetListTableViewController.swift (hunk 2 above)
Comment (udayasri) [reply RefApp/ViewControllers/AssetListTableViewController.swift]: I have considered to use pop when logging out, but as we will have two ViewControllers (EnvironmentViewController & LoginViewController ) in the stack we may need to use something like this  `popToViewController(self.navigationController?.viewControllers[0], , animated: true)` to directly navigate to  `EnvironmentViewController`. 

Instead that I used  dismissing the `LoginViewController` & making the `AssetListTableViewController` the rootView when logging the user. Let me know if you have more suggestions.

**Pull request merged**